Assets/Tests/PlayModeTests/DeviceIdTests.cs

```diff
@@ -377,22 +379,100 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
+            {
+                ValidateDeviceIDAndType(Countly.Instance, id, expectedIdType, true);
+            }
+            else
+            {
+                ValidateDeviceIDAndType(Countly.Instance, deviceID, expectedIdType, true);
+            }
+        }
+
+        // 'SetId' method in DeviceIdCountlyService
+        // We provide empty, null, whitespace and valid ids to SetId method over SDKGenerated id
+        // Id should only change in case of a valid id
+        [TestCase("", DeviceIdType.SDKGenerated, false)]
+        [TestCase(null, DeviceIdType.SDKGenerated, false)]
+        [TestCase(" ", DeviceIdType.SDKGenerated, false)]
+        [TestCase("new_user", DeviceIdType.DeveloperProvided, true)]
+        public void SetId_OverSDKGenerated(string id, DeviceIdType expectedIdType, bool isExpectingChange)
+        {
+            CountlyConfiguration config = new CountlyConfiguration(_appKey, _serverUrl);
+            Countly.Instance.Init(config);
+            string deviceID = Countly.Instance.Device.DeviceId; // sdk generated device id
+
+            ValidateDeviceIDAndType(Countly.Instance, deviceID, DeviceIdType.SDKGenerated, true);
+            Countly.Instance.Device.SetId(id);
+            if (isExpectingChange)
+            {
+                ValidateDeviceIDAndType(Countly.Instance, id, expectedIdType, true);
+            }
+            else
+            {
+                ValidateDeviceIDAndType(Countly.Instance, deviceID, expectedIdType, true);
+            }
+        }
+
+        // 'SetId' method in DeviceIdCountlyService
+        // We provide the same device id that SDK generated 
+        // Id should remain same and type shouldn't change
+        [TestCase(DeviceIdType.SDKGenerated)]
+        [TestCase(DeviceIdType.DeveloperProvided)]
+        public void SetId_SameId(DeviceIdType startType)
+        {
+            string deviceId = "";
+            if (startType == DeviceIdType.SDKGenerated)
+            {
+                CountlyConfiguration config = new CountlyConfiguration(_appKey, _serverUrl);
+                Countly.Instance.Init(config);
+                deviceId = Countly.Instance.Device.DeviceId; // sdk generated device id
+
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.SDKGenerated, true);
+
+                Countly.Instance.Device.SetId(deviceId);
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.SDKGenerated, true);
+            }
+            else
+            {
+                Countly.Instance.Init(TestUtility.CreateBaseConfig());
+                deviceId = Countly.Instance.Device.DeviceId; // "test_user"
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.DeveloperProvided, true);
+
+                Countly.Instance.Device.SetId(deviceId);
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.DeveloperProvided, true);
+            }
+        }
+
```

Comment:
can you add 1 more test where you double call setid with different values and verify at each step id is correct